src/charm.py

```diff
@@ -24,36 +29,73 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.irc = IRCService()
+        self._irc = IRCBridgeService()
+        self._database = DatabaseObserver(self, DATABASE_RELATION_NAME)
+        self._matrix = MatrixObserver(self, MATRIX_RELATION_NAME)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.stop, self._on_stop)
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
 
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration."""
-        self.unit.status = ops.ActiveStatus()
+        """Handle config changed."""
+        self.reconcile()
 
     def _on_install(self, _: ops.InstallEvent) -> None:
         """Handle install."""
-        self.unit.status = ops.MaintenanceStatus("Preparing irc bridge")
-        self.irc.prepare()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
 
     def _on_start(self, _: ops.StartEvent) -> None:
         """Handle start."""
-        self.irc.start()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
 
     def _on_stop(self, _: ops.StopEvent) -> None:
         """Handle stop."""
-        self.irc.stop()
+        self._irc.stop()
+
+    @property
```

Comment:
Since CharmConfig might return an exception, either you manage the exception here or the method shouldn't a property

Reply:
Decided to remove the property decorator, thanks!